### ISSUE TITLE
gep: how to handle vector types?

### ISSUE THREAD

**Issue by mewmew** [llir/llvm#107]
I'm wondering exactly how the `gep` instructions semantics are to be interpreted when passed a vector argument. How does it compute the result type?

From the updated test cases of LLVM 9.0, we have one which uses `getelementptr` with a vector as argument.

```llvm
define i32* @gep_vbase_w_s_idx(<2 x i32*> %base) {
; CHECK-LABEL: @gep_vbase_w_s_idx(
; CHECK-NEXT:    [[GEP:%.*]] = getelementptr i32, <2 x i32*> [[BASE:%.*]], i64 1
; CHECK-NEXT:    [[EE:%.*]] = extractelement <2 x i32*> [[GEP]], i32 1
; CHECK-NEXT:    ret i32* [[EE]]
;
  %gep = getelementptr i32, <2 x i32*> %base, i64 1
  %ee = extractelement <2 x i32*> %gep, i32 1
  ret i32* %ee
}
```

Why is the type of `%gep` a vector type (i.e. `<2 x i32*>`) and not a pointer to integer type (i.e. (`i32*`)? This is what I can't figure out.

Anyone who feels like helping me gain clarity? It would be much appreciated.

For background, this currently causes a failure for the following test case:

```diff
2019/11/01 16:09:24 === [ ../testdata/llvm/test/Transforms/InstCombine/vec_demanded_elts.ll ] ===
diff --git a/vec_demanded_elts.ll.golden b/vec_demanded_elts.ll.golden
index ac66f6a..729daa1 100644
--- a/vec_demanded_elts.ll.golden
+++ b/vec_demanded_elts.ll.golden
@@ -328,7 +328,7 @@ define <3 x float> @shuf_frem_const_op1(<3 x float> %x) {
 define i32* @gep_vbase_w_s_idx(<2 x i32*> %base) {
 ; <label>:0
 	%gep = getelementptr i32, <2 x i32*> %base, i64 1
-	%ee = extractelement <2 x i32*> %gep, i32 1
+	%ee = extractelement i32* %gep, i32 1
 	ret i32* %ee
 }
 
@@ -337,7 +337,7 @@ define i32* @gep_splat_base_w_s_idx(i32* %base) {
 	%basevec1 = insertelement <2 x i32*> undef, i32* %base, i32 0
 	%basevec2 = shufflevector <2 x i32*> %basevec1, <2 x i32*> undef, <2 x i32> zeroinitializer
 	%gep = getelementptr i32, <2 x i32*> %basevec2, i64 1
-	%ee = extractelement <2 x i32*> %gep, i32 1
+	%ee = extractelement i32* %gep, i32 1
 	ret i32* %ee
 }
 
@@ -362,7 +362,7 @@ define i32* @gep_splat_base_w_vidx(i32* %base, <2 x i64> %idxvec) {
 define i32* @gep_cvbase_w_s_idx(<2 x i32*> %base, i64 %raw_addr) {
 ; <label>:0
 	%gep = getelementptr i32, <2 x i32*> <i32* @GLOBAL, i32* @GLOBAL>, i64 %raw_addr
-	%ee = extractelement <2 x i32*> %gep, i32 1
+	%ee = extractelement i32* %gep, i32 1
 	ret i32* %ee
 }
```

**Comment by dannypsnl** [comment]
I'm not pretty sure what does [this](https://llvm.org/docs/GetElementPtr.html#can-gep-index-into-vector-elements) means. But it looks like a misuse for gep?

According:

> and fundamental inconsistency in the IR.

**Comment by mewmew** [comment]
Thanks @dannypsnl! Then that's definitely the direction that llir/llvm would be heading as well.

From https://llvm.org/docs/GetElementPtr.html#can-gep-index-into-vector-elements
> **Can GEP index into vector elements?**
>
> In the future, it will probably be outright disallowed.

**Comment by dannypsnl** [comment]
IMO, we can disallow it at here? Since the behavior is look like an undefined behavior in LLVM.

**Comment by mewmew** [comment]
> IMO, we can disallow it at here? Since the behavior looks like undefined behavior in LLVM.

@dannypsnl, I think we should disallow it when LLVM officially disallows it. For now, it's enough to leave as is and just turn off the test case. Note that prior to the LLVM 9.0 test case, it was passing, and since the behaviour is undefined as per gep FAQ, we can just leave it as is.

**Comment by mewmew** [comment]
"fixed" in 3c4d1a35ba23daf6e161d060feffadfc518a5215. We will disallow the use of gep for indexing into vectors of pointers entirely in the future, when LLVM decides to do so.
